### PR TITLE
Add fancy mode for diagnostics/errors

### PR DESCRIPTION
**Pull request by viniciusmuller**
This pull request adds support for fancier and friendlier error/warning messages, as discussed in #12644.

To see how it currently looks, you can take a look at the tests.

It also plays nice with non-elixir files, such as phoenix templates:
image

### CODE REVIEW

File: lib/elixir/src/elixir_errors.erl [hunk 1]
```diff
@@ -22,18 +24,234 @@ print_warning(Location, File, Message) ->
...
+ {match, [MatchingTerm]} = re:run(Tail, Re, [{capture, all, binary}]),
+
+  case MatchingTerm of 
+    % If we cannot consume anything, return a single ^ at column position
+    <<>> -> 1;
+    _ -> string:length(MatchingTerm)
+  end.
+
+get_file_line(File, LineNumber) -> 
+  {ok, Data} = file:read_file(File),
```
Comment (viniciusmuller) [inline lib/elixir/src/elixir_errors.erl:239]: Here we are reading the whole file just to get the line, which is not ideal, I was using this code snippet before:

```erl
  {ok, IoDevice} = file:open(File, [read]),
  LineCollector = fun 
                    (I, nil) when I == LineNumber - 1 -> 
                      {ok, Line} = file:read_line(IoDevice),
                      string:trim(Line, trailing);
                    (_, nil) -> 
                      {ok, _} = file:read_line(IoDevice),
                      nil;
                    (_, Line) -> 
                      Line
                  end,
  Line = lists:foldl(LineCollector, nil, lists:seq(0, LineNumber)),
  file:close(IoDevice),
  Line.
```

But I couldn't get it working with unicode after a lot of adaptations, so ideally we should make it handle unicode and use it.

**Comment by josevalim** [comment]
@viniciusmuller this looks great! :heart:

Two nits:

1. Should we make the line be a bit longer?


```elixir
  |- error: ...
  |
  | snippet
  |        ^
  |
  message
  
```

The reason for this is because otherwise it is not clear if the warning belongs to the message before or after (especially if you have many of them).

2. The pull request is currently large. Can you please send a PR that first changes the code snippet handling for SyntaxError and TokenMissingError, since those already handle snippet logic, and then we generalize that? Thank you!

I'm super happy with your contributions here!

**Comment by viniciusmuller** [comment]
> Should we make the vertical line be a bit longer?

I think this makes sense, I've also felt the messages were a bit "too separated".

> The pull request is currently large. Can you please send a PR that first changes the code snippet handling for SyntaxError and TokenMissingError, since those already handle snippet logic, and then we generalize that? Thank you!

Yep, we can do that!

> I'm super happy with your contributions here!

Thanks for the feedback! <3

**Comment by viniciusmuller** [comment]
Closing in favor of #12670

**Pull request closed without merging**